api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,183 @@
+package protocol
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"reflect"
+	"time"
+
+	"go.opencensus.io/trace"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+// CloudEvent is the official JSON representation of a CloudEvent: https://github.com/cloudevents/spec/blob/master/serialization.md
+type CloudEvent struct {
+	CloudEventsVersion string                 `json:"cloudEventsVersion"`
+	EventID            string                 `json:"eventID"`
+	Source             string                 `json:"source"`
+	EventType          string                 `json:"eventType"`
+	EventTypeVersion   string                 `json:"eventTypeVersion"`
+	EventTime          time.Time              `json:"eventTime"` // TODO: ensure rfc3339 format
+	SchemaURL          string                 `json:"schemaURL"`
```

Comment:
has to be a URL to data schema according to the spec https://github.com/cloudevents/spec/blob/master/spec.md#schemaurl

Reply:
that's just a string though, at least https://golang.org/pkg/net/url/#URL doesn't have UnmarshalJSON so we need a custom type (from non-stdlib package or our own)

Reply:
we should validate a few of these, for sure*